039-user-operator-custom-filter-for-labels-.md

```diff
@@ -0,0 +1,139 @@
+  name: my-user
+  namespace: myproject
+...
+```
+We can see that User Operator automatically adds the `app.kubernetes.io/instance: my-user` label.
+In some situations, we do not want a such label to be automatically created (i.e., [endless deletion and creation of specific Secret](https://github.com/strimzi/strimzi-kafka-operator/issues/5690))
+and therefore, one way how to get rid of this already assigned label is to filter it.
+
+## Proposal
+
+This proposal suggests adding configurable exclusion of labels.
+We can implement this feature similarly to [issue](https://github.com/strimzi/strimzi-kafka-operator/pull/4791).
+We can create an environment variable and inject such value through the `Kafka` custom resource.
+Specifically, in the `spec.entityOperator.template.topicOperatorContainer.env`.
+For instance, we can have the following `Kafka` resource:
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  ...
+  entityOperator:
+    template:
+      userOperatorContainer:
+        env:
+          name:   STRIMZI_LABELS_EXCLUSION_PATTERN
+          value:  "^app.kubernetes.io/.*$"
+    topicOperator: {}
+    userOperator: {}
+```
+Then, we would obtain a value from the environment variable and parse it in the `UserOperatorConfig` class.
+Based on the environment variable value, we will determine what to do (a) env is not specified, and therefore we do not exclude any label from Secret
+(b) env is specified by a regex, and then we will filter and remove such labels which match the input regex.
+
+### More implementation details
+
+### 1 Parsing part
+
+As we mentioned details about the `UserOperatorConfig` class, we will elaborate more on the specific details in this section.
+Moreover, we will obtain the environment variable inside the `fromMap` method used to construct the class.
+```java
+public static UserOperatorConfig fromMap(Map<String, String> map) {
+    ...
+    // 
+    String strimziLabelsExclusionPatternEnvVar = map.get(UserOperatorConfig.STRIMZI_LABELS_EXCLUSION_PATTERN);
+    
+    if (strimziLabelsExclusionPattern != null) {
+        strimziLabelsExclusionPattern = strimziLabelsExclusionPatternEnvVar;
+    }
+    ...
+}
+```
+
+### 2 Exclusion part
+
+The exclusion part will be placed in the `KafkaUserModel` class.
+Moreover, we would need to get the value of the environment variable to such a class.
+In the `UserOperatorConfig` class, we will obtain value, and then in the `KafkaUserOperaor` class, specifically in the following method:
+```java
+protected Future<KafkaUserStatus> createOrUpdate(Reconciliation reconciliation, KafkaUser resource) {
+    KafkaUserModel user;
+    KafkaUserStatus userStatus = new KafkaUserStatus();
+
+    try {
+        user = KafkaUserModel.fromCrd(resource, config.getSecretPrefix(), config.isAclsAdminApiSupported(),config.isKraftEnabled(), 
+            config.getStrimziLabelsExclusionPattern); // <-- this one we will inject into KafkaUserModel class
```

Comment:
This way, you will ahve to compile the variable again and again in every reocnciliation. That seems like waste of resources.

Reply:
I have changed the String type of strimziLabelsExclusionPattern to the `Pattern` and also compile such regex in the parsing part (i.e., in the UserOperatorConfig), which should eliminate this workload...